evolver/history/interface.py

```diff
@@ -0,0 +1,54 @@
+from abc import abstractmethod
+from typing import Any
+
+from pydantic import BaseModel
+
+from evolver.base import BaseInterface
+
+
+class HistoricDatum(BaseModel):
+    timestamp: float
+    data: Any
+
+
+class HistoryResult(BaseModel):
+    data: dict[str, list[HistoricDatum]]
+
+
+class History(BaseInterface):
+    class Config(BaseInterface.Config):
+        pass
+
+    @abstractmethod
+    def put(self, name: str, data: Any):
+        """Add data for hardware component to history.
+
+        Args:
+            name: The name of hardware component to save data for.
+            data: The data to save. This MUST be a JSON-serializable object, as
+                per the FastAPI json serialization requirements.
+        """
+        pass
+
+    @abstractmethod
+    def get(self, name: str = None, t_start: float = None, t_stop: float = None, n_max: int = None) -> HistoryResult:
+        """Get saved history data.
+
+        Implementations must override this method and return a HistoryResult object
+        according to the input parameters, as listed below.
+
+        Args:
+            name: The name of hardware component to retrieve data for. If None,
+                return all data.
+            t_start: The start time of the data to retrieve in floating point unix
+                epoch seconds. If not specified, either n_max or an
+                implementation-defined number of data points will be returned,
+                prior to t_stop or now.
+            t_stop: The stop time of the data to retrieve in floating point unix
+                epoch seconds. If not specified, either n_max or an
+                implementation-defined numer of data points will be returned,
+                after t_start or up to now.
+            n_max: The maximum number of data points to retrieve. If unspecified,
```

Comment:
1.) seems a good idea, the `HistoryResult` structure could be populated with that, which I suppose is most useful as a per-name (per hardware) basis

2.) this is definitely targeting time-based views, which is arguably the primary objective of the history service. I wonder if views that abstract over time, (I'm thinking like scatter of temp vs od) would be better served in a more flexible but less formalized query endpoint. Such an endpoint would be more general, but maybe be more onerous for plotting the most common (time-domain) plots.

2b.) This might be not-well defined, since we likely expect fresh data to come in any time - would an empty result suffice as an indicator?

Reply:
That all makes sense, i think 2b is addressed if 1. is adopted, since the "end of results" will be supplied by that value, and yes, "the time right now" provides the other bound.

Yes that's it exactly for 2, some tradeoffs being made.